massa-async-pool/src/pool.rs

```diff
@@ -130,76 +138,153 @@ impl AsyncPool {
             .collect()
     }
 
-    /// Used for bootstrap
-    /// Take a part of the async pool starting from the next element after `last_id` and with a max length of the constant `ASYNC_POOL_PART_SIZE_MESSAGE_BYTES`.
-    /// Should always follow the same behavior as the `get_ledger_part` from `FinalLedger`method.
+    /// Get a part of the async pool.
+    /// Used for bootstrap.
+    ///
+    /// # Arguments
+    /// * cursor: current bootstrap state
+    ///
+    /// # Returns
+    /// The async pool part and the updated cursor
     pub fn get_pool_part(
         &self,
-        last_id: Option<AsyncMessageId>,
-    ) -> Result<(Vec<u8>, Option<AsyncMessageId>), ModelsError> {
-        let last_id = if let Some(last_id) = last_id {
-            Excluded(last_id)
-        } else if self.messages.first_key_value().is_some() {
-            Unbounded
-        } else {
-            return Ok((Vec::new(), None));
+        cursor: StreamingStep<AsyncMessageId>,
+    ) -> Result<
+        (
+            BTreeMap<AsyncMessageId, AsyncMessage>,
+            StreamingStep<AsyncMessageId>,
+        ),
+        ModelsError,
+    > {
+        let mut pool_part = BTreeMap::new();
+        let left_bound = match cursor {
+            StreamingStep::Started => Unbounded,
+            StreamingStep::Ongoing(last_id) => Excluded(last_id),
+            StreamingStep::Finished => return Ok((pool_part, cursor)),
         };
-        let mut part = Vec::new();
-        let mut next_last_id = None;
-        let id_async_message_serializer = AsyncMessageIdSerializer::new();
-        let async_message_serializer = AsyncMessageSerializer::new();
-        for (id, message) in self.messages.range((last_id, Unbounded)) {
-            if part.len() < self.config.part_size_message_bytes as usize {
-                id_async_message_serializer.serialize(id, &mut part)?;
-                async_message_serializer.serialize(message, &mut part)?;
-                next_last_id = Some(*id);
+        for (id, message) in self.messages.range((left_bound, Unbounded)) {
+            if pool_part.len() < self.config.bootstrap_part_size as usize {
+                pool_part.insert(*id, message.clone());
             }
         }
-        Ok((part, next_last_id))
+        if pool_part.is_empty() {
+            return Ok((pool_part, StreamingStep::Finished));
+        };
+        let pool_part_last_id = pool_part
```

Comment:
Isn't it more optimize to keep a save of the last element inserted in your loop above instead of this ? Maybe it's not because it's a `BTreeMap` but not sure